primitives/blockchain/src/error.rs

```diff
@@ -28,121 +28,175 @@ pub type Result<T> = result::Result<T, Error>;
-	RuntimeParamConversion(String),
-	/// Changes tries are not supported.
+
+	#[error("Decoding of {0} failed")]
+	RuntimeApiCodecError(&'static str, #[source] codec::Error),
+
+	#[error("Runtime :code missing in storage")]
+	RuntimeCodeMissing,
+
 	#[error("Changes tries are not supported by the runtime")]
 	ChangesTriesNotSupported,
-	/// Error reading changes tries configuration.
+
 	#[error("Error reading changes tries configuration")]
 	ErrorReadingChangesTriesConfig,
-	/// Key changes query has failed.
+
 	#[error("Failed to check changes proof: {0}")]
 	ChangesTrieAccessFailed(String),
-	/// Last finalized block not parent of current.
+
 	#[error("Did not finalize blocks in sequential order.")]
 	NonSequentialFinalization(String),
-	/// Safety violation: new best block not descendent of last finalized.
+
 	#[error("Potential long-range attack: block not in finalized chain.")]
 	NotInFinalizedChain,
-	/// Hash that is required for building CHT is missing.
+
 	#[error("Failed to get hash of block for building CHT")]
 	MissingHashRequiredForCHT,
-	/// Invalid calculated state root on block import.
+
 	#[error("Calculated state root does not match.")]
 	InvalidStateRoot,
-	/// Incomplete block import pipeline.
+
 	#[error("Incomplete block import pipeline.")]
 	IncompletePipeline,
+
 	#[error("Transaction pool not ready for block production.")]
 	TransactionPoolNotReady,
+
 	#[error("Database")]
 	DatabaseError(#[from] sp_database::error::DatabaseError),
-	/// A convenience variant for String
-	#[error("{0}")]
-	Msg(String),
-}
 
-impl<'a> From<&'a str> for Error {
-	fn from(s: &'a str) -> Self {
-		Error::Msg(s.into())
-	}
-}
+	#[error("Failed to get header for hash {0}")]
+	MissingHeader(String),
 
-impl From<String> for Error {
-	fn from(s: String) -> Self {
-		Error::Msg(s)
-	}
+	#[error("Failed to load the block weight for block {0}")]
+	LoadingBlockWeightFailed(String),
+
+	#[error("WASM override IO error")]
+	WasmOverrideIo(PathBuf, #[source] std::io::Error),
+
+	#[error("Overwriting WASM requires a directory where local \
+	WASM is stored. {} is not a directory", .0.display())]
+	WasmOverrideNotADirectory(PathBuf),
+
+	#[error("Duplicate WASM Runtimes found: \n{}\n", .0.join("\n") )]
+	DuplicateWasmRuntime(Vec<String>),
+
+	#[error("State Database error: {0}")]
+	StateDatabase(String),
+
+	#[error("Invalid value for slot_duration: the value ({0}) must be greater than 0.")]
+	SlotDurationInvalid(u64),
+
+	#[error(transparent)]
+	Application(#[from] Box<dyn std::error::Error + Send + Sync + 'static>),
+
+	#[error("AlwaysBadChecker")]
+	AlwaysBadChecker,
+
+	// Should be removed/improved once
+	// the storage `fn`s returns typed errors.
+	#[error("Runtime code error: {0}")]
+	RuntimeCode(&'static str),
+
+	// Should be removed/improved once
+	// the storage `fn`s returns typed errors.
+	#[error("Storage error: {0}")]
+	Storage(String),
+	
+	/// Unit test helper for mocked modules.
+	#[error("Mock: {0}")]
+	Mock(&'static str),
```

Comment:
Use a local type in a unit test.

I don't want to see testing related error variants in the normal error type.